### PR TITLE
Ensure @BeforeGroups gets called only once

### LINKED ISSUE #1108: @BeforeGroups called twice
```java
public class testNG {

    @BeforeGroups(groups = {"smoketests", "functionaltests"})
    public void before() {
        System.out.println("Before Groups");
    }

    @Test(groups = {"smoketests", "functionaltests"})
    public void test() {
        System.out.println("Test");
    }

    @AfterGroups(groups = {"smoketests", "functionaltests"})
    public void after() {
        System.out.println("After Groups");
    }
}
```

Will display:

```
Before Groups
Before Groups
Test
After Groups
```

From: http://stackoverflow.com/q/38796087/4234729

### PR DESCRIPTION
**Pull request by krmahadevan**
Fixes #1108

### CODE REVIEW

**Comment by krmahadevan** [comment]
@juherr Can you please help with reviewing this ?

File: src/main/java/org/testng/internal/Invoker.java [hunk 1]
```diff
@@ -1609,6 +1609,21 @@ public boolean isTrue(ITestNGMethod m, IClass c) {
     return result;
   }
 
+  private boolean isConfiguration(ITestNGMethod method) {
```
Comment (juherr) [inline src/main/java/org/testng/internal/Invoker.java]: Should be static, and should be moved into `BaseTestMethod` or in an `TestNgMethodUtils`-like class

Comment (juherr) [inline src/main/java/org/testng/internal/Invoker.java]: Take the time to use this method for `InvokedMethod#isConfigurationMethod` too.

Then, rename the method to `isConfigurationMethod`.

File: src/main/java/org/testng/internal/Invoker.java [hunk 2]
```diff
@@ -1609,6 +1609,21 @@ public boolean isTrue(ITestNGMethod m, IClass c) {
...
+    return method.isAfterClassConfiguration() || method.isBeforeClassConfiguration() ||
+        method.isBeforeGroupsConfiguration() || method.isAfterGroupsConfiguration() ||
+        method.isBeforeMethodConfiguration() || method.isAfterMethodConfiguration() ||
+        method.isBeforeTestConfiguration() || method.isAfterTestConfiguration() ||
+        method.isBeforeSuiteConfiguration() || method.isAfterSuiteConfiguration();
+  }
+
+  private boolean alreadyAdded(ITestNGMethod method, List<ITestNGMethod> methods) {
```
Comment (juherr) [inline src/main/java/org/testng/internal/Invoker.java]: Same: static + location.

About the naming, could be `containsConfigurationMethod(List<> methods, ITest.. method)`

File: src/test/java/org/testng/internal/invokers/InvokerTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Test
+    public void testClassWithUniqueGroups() {
+        valueMap.clear();
+        privateRun(ClassWithUniqueGroupNames.class, SMOKE);
+        Assert.assertEquals(new Integer(1), valueMap.get(ClassWithUniqueGroupNames.class.getSimpleName()));
+    }
+
+    private void privateRun(final Class className, String... groupNames) {
```
Comment (juherr) [inline src/test/java/org/testng/internal/invokers/InvokerTest.java]: `Class<?>`

File: src/test/java/org/testng/internal/invokers/InvokerTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        final XmlSuite suite = new XmlSuite();
+        suite.setName("simple-suite");
+        final XmlTest xmlTest = new XmlTest(suite);
+        xmlTest.setName("simple-test");
+        xmlTest.setClasses(new ArrayList<XmlClass>() {{
```
Comment (juherr) [inline src/test/java/org/testng/internal/invokers/InvokerTest.java]: Prefer `Collections#singletonList` or `Arrays#asList`

File: src/test/java/org/testng/internal/invokers/InvokerTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            add(new XmlClass(className));
+        }});
+        for (String group : groupNames) {
+            xmlTest.addIncludedGroup(group);
+
+        }
+        suite.setTests(new ArrayList<XmlTest>() {{
+            add(xmlTest);
+        }});
+        TestNG tng = create();
```
Comment (juherr) [inline src/test/java/org/testng/internal/invokers/InvokerTest.java]: Use `create(XmlSuite)`

File: src/test/java/org/testng/internal/invokers/InvokerTest.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        tng.setXmlSuites(new ArrayList<XmlSuite>() {{
+            add(suite);
+        }});
+        tng.run();
+    }
+
+    static class LocalBase {
+        protected void before(String name) {
+            Map<String, Integer> map = InvokerTest.getValueMap();
```
Comment (juherr) [inline src/test/java/org/testng/internal/invokers/InvokerTest.java]: You can avoid such complexity by using the `InvokedMethodNameListener` and checking values.

File: src/test/java/org/testng/internal/invokers/InvokerTest.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                value = 1;
+            } else {
+                value++;
+            }
+            map.put(name, value);
+        }
+    }
+
+
+    public static class ClassWithUniqueGroupNames extends LocalBase {
```
Comment (juherr) [inline src/test/java/org/testng/internal/invokers/InvokerTest.java]: Suffix sample with `*Sample`. The convention is `*Test` for tests and `*Sample` for their samples.

File: src/test/java/org/testng/internal/invokers/InvokerTest.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
+package org.testng.internal.invokers;
```
Comment (juherr) [inline src/test/java/org/testng/internal/invokers/InvokerTest.java]: Use `test.groupinvocation` package instead

File: src/test/java/org/testng/internal/invokers/InvokerTest.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    @Test
+    public void testClassWithUniqueGroups() {
+        valueMap.clear();
+        privateRun(ClassWithUniqueGroupNames.class, SMOKE);
+        Assert.assertEquals(new Integer(1), valueMap.get(ClassWithUniqueGroupNames.class.getSimpleName()));
+    }
+
+    private void privateRun(final Class className, String... groupNames) {
+        final XmlSuite suite = new XmlSuite();
+        suite.setName("simple-suite");
```
Comment (juherr) [inline src/test/java/org/testng/internal/invokers/InvokerTest.java]: `SimpleBaseTest` provides some util methods `createXmlSuite`, `createXmlTest`, ...

**Comment by juherr** [comment]
LGTM after the little changes on form :+1:

**Comment by krmahadevan** [comment]
@juherr Thanks for taking the time to help with the review. I have fixed all the review comments. Please help with merge.

**Comment by juherr** [comment]
:+1: 

Ping @cbeust

File: src/main/java/org/testng/internal/TestNgMethodUtils.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        //Utility class. So hiding the constructor.
+    }
+
+    /**
+     * A helper method that checks to see if a method is a configuration method or not.
+     * @param method - A {@link ITestNGMethod} object which needs to be checked.
+     * @return - <code>true</code> if the method is a configuration method and false if its a test method.
+     */
+    static boolean isConfigurationMethod(ITestNGMethod method) {
+        return isConfigurationMethod(method, false);
```
Comment (cbeust) [inline src/main/java/org/testng/internal/TestNgMethodUtils.java]: Nit pick: I like to explain what magical boolean values do:

```
 return isConfigurationMethod(method, false /* don't include group configs */);
```

**Pull request merged**